### ISSUE TITLE
proxy error on execution

### ISSUE THREAD

**Issue by choosehappy** [camicroscope/Distro#65]
I've issued these commands:

```
git clone https://github.com/camicroscope/Distro.git
cd Distro/
docker-compose -f caMicroscope.yml up
```

which appear to work okay. This leaves me with this:

![image](https://user-images.githubusercontent.com/9681868/63608623-bfe9fe80-c5d5-11e9-8933-d401d7719810.png)

Everything look good so far

Then when i go to: http://haeckel.case.edu:4010/apps/loader/loader.html

I see a landing page, cool

clicking on load image page, i select a local image, and click upload, but see this:

```
ca-security | {"__statusCode":400,"err":"BAD REQUEST","type":"proxy error"}
ca-security | { Error: write ECONNRESET
ca-security |     at WriteWrap.afterWrite [as oncomplete] (net.js:868:14) errno: 'ECONNRESET', code: 'ECONNRESET', syscall: 'write' }
ca-security | [HPM] Error occurred while trying to proxy request upload/continue/KOY25R9L4D from haeckel.case.edu:4010 to http://ca-load:4000 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
ca-security | Error: write ECONNRESET
ca-security |     at WriteWrap.afterWrite [as oncomplete] (net.js:868:14)
ca-security | public check false
ca-security | [HPM] Proxy created: /  ->  http://ca-load:4000
ca-security | 6990556
ca-security | {"__statusCode":400,"err":"BAD REQUEST","type":"proxy error"}
ca-security | { Error: write ECONNRESET
ca-security |     at WriteWrap.afterWrite [as oncomplete] (net.js:868:14) errno: 'ECONNRESET', code: 'ECONNRESET', syscall: 'write' }
ca-security | [HPM] Error occurred while trying to proxy request upload/continue/KOY25R9L4D from haeckel.case.edu:4010 to http://ca-load:4000 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
ca-security | Error: write ECONNRESET
ca-security |     at WriteWrap.afterWrite [as oncomplete] (net.js:868:14)
ca-security | public check false
```

A number of times. No file is ultimately uploaded

going to:

http://haeckel.case.edu:4010/apps/table.html

produces the message "No data Found .... x _x"

I have confirmed this exact behavior on 2 different machines. both are ubuntu 16.04

Any ideas?

**Comment by birm** [comment]
Hello, and thanks for using caMicroscope!
So we can see what may be wrong, can you please turn on logs for the loader container as well? This is accomplished by deleting lines 54 and 55 of https://github.com/camicroscope/Distro/blob/de335e7a0207b9ab012e9c3fe339d46b9641c916/caMicroscope.yml#L54

Thanks again,
-Ryan

**Comment by choosehappy** [comment]
yes sir


so i've done that but it doesn't seem to have produced any more output?

i've removed those 2 lines:

![image](https://user-images.githubusercontent.com/9681868/63611593-7ea91d00-c5dc-11e9-90be-2f106ffacbc8.png)

and then rerun the up command:
```
root@haeckel:~/Distro# /usr/local/bin/docker-compose -f caMicroscope.yml up
Starting ca-front    ... done
Starting ca-mongo    ... done
Starting ca-iip      ... done
Starting ca-security ... done
Starting ca-load     ... done
Starting ca-auth     ... done
Starting ca-idx      ... done
Starting ca-data     ... done
Attaching to ca-mongo, ca-auth, ca-front, ca-iip, ca-security, ca-load, ca-data, ca-idx
ca-mongo    | WARNING: no logs are available with the 'none' log driver
ca-auth     | WARNING: no logs are available with the 'none' log driver
ca-front    | WARNING: no logs are available with the 'none' log driver
ca-data     | WARNING: no logs are available with the 'none' log driver
ca-idx      | WARNING: no logs are available with the 'none' log driver
ca-security | warn:    --minUptime not set. Defaulting to: 1000ms
ca-security | warn:    --spinSleepTime not set. Your script will exit if it does not stay up for at least 1000ms
ca-load     | [2019-08-23 17:30:58 +0000] [6] [INFO] Starting gunicorn 19.9.0
ca-load     | [2019-08-23 17:30:58 +0000] [6] [INFO] Listening at: http://0.0.0.0:4000 (6)
ca-load     | [2019-08-23 17:30:58 +0000] [6] [INFO] Using worker: sync
ca-load     | [2019-08-23 17:30:58 +0000] [9] [INFO] Booting worker with pid: 9
ca-load     | [2019-08-23 17:30:58 +0000] [10] [INFO] Booting worker with pid: 10
ca-load     | [2019-08-23 17:30:58 +0000] [11] [INFO] Booting worker with pid: 11
ca-load     | [2019-08-23 17:30:58 +0000] [12] [INFO] Booting worker with pid: 12
ca-security | listening on 4010
ca-idx exited with code 0
ca-security | public check true
ca-security | [HPM] Proxy created: /  ->  http://ca-front:80
ca-security | public check true
ca-security | [HPM] Proxy created: /  ->  http://ca-front:80
ca-security | public check false
ca-security | [HPM] Proxy created: /  ->  http://ca-load:4000
ca-security | 34
ca-security | public check false
ca-security | [HPM] Proxy created: /  ->  http://ca-load:4000
ca-security | 6990550
ca-security | public check false
ca-security | [HPM] Proxy created: /  ->  http://ca-load:4000
ca-security | 34
ca-security | {"__statusCode":400,"err":"BAD REQUEST","type":"proxy error"}
ca-security | public check false
ca-security | [HPM] Proxy created: /  ->  http://ca-load:4000
ca-security | 6990556
ca-security | public check false
ca-security | [HPM] Proxy created: /  ->  http://ca-load:4000

```

**Comment by choosehappy** [comment]
are the logs to the screen or to a file? doesn't look like a file was created:

```
axj232@haeckel:~/Distro$ find . | grep log
./db/_mdb_catalog.wt
./db/journal/WiredTigerPreplog.0000000002
./db/journal/WiredTigerPreplog.0000000001
./.git/logs
./.git/logs/refs
./.git/logs/refs/heads
./.git/logs/refs/heads/master
./.git/logs/refs/remotes
./.git/logs/refs/remotes/origin
./.git/logs/refs/remotes/origin/HEAD
./.git/logs/HEAD
./auth0_login.html
./config/login.html
./login.html
```

**Comment by choosehappy** [comment]
by the way, the UI also produces the same error message when the file is eventually (fake) uploaded :)

![image](https://user-images.githubusercontent.com/9681868/63612217-03e10180-c5de-11e9-8dc1-b2b12f173239.png)

**Comment by birm** [comment]
To confirm; this error occurs when you choose a file, or when you click "finish upload"?
In either case, it's probably not a bad idea to get your client network logs from the loader ui page.

Thank you again,
-Ryan

**Comment by choosehappy** [comment]
it definitely occurs after clicking "finish upload"

"it's probably not a bad idea to get your client network logs from the loader ui page."

sorry, i'm not sure what this means, i don't see any option for logs?

**Comment by birm** [comment]
Ah, my apologies - this is part of the developer console; for firefox see https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor
for chrome, see https://developers.google.com/web/tools/chrome-devtools/network/

**Comment by choosehappy** [comment]
ah, yes, that one. are you looking for the full har?

https://www.dropbox.com/s/qpoenhwp5b3c6kh/haeckel.case.edu.har?dl=0


Nothing else of interest appears in the console:

![image](https://user-images.githubusercontent.com/9681868/63613943-2117cf00-c5e2-11e9-9245-8d331aa7c840.png)

**Comment by birm** [comment]
what happens when you "check" that filename?

**Comment by choosehappy** [comment]
![image](https://user-images.githubusercontent.com/9681868/63614324-072abc00-c5e3-11e9-94cb-6da18e53fb2d.png)


https://www.dropbox.com/s/yqbrf3ixa3lqtj6/har_check.txt?dl=0

**Comment by birm** [comment]
do you get anything with this url? http://haeckel.case.edu:4010/load/Slide/thumb/B15_4436_4_CETL.svs

also, you could check the images directory on the server (probably "images" with in the distro repository) to see if the file exists there.

Thank you for your patience,
-Ryan

**Comment by choosehappy** [comment]
i've tried with a number of files,:

```
-rw-r--r-- 1 root   root          0 Aug 23 12:39 B13_7861_1_CETL.svs
-rw-r--r-- 1 root   root   47185920 Aug 23 12:47 B15_4436_4_CETL.svs
-rw-r--r-- 1 root   root   10485760 Aug 23 13:29 B17_6663_MCT.svs
-rw-r--r-- 1 root   root          0 Aug 23 14:42 PD10011a_HE-2014-12-07_15.02.40.svs
```

it doesn't look like any of them have been fully uploaded


i guess as expected when going to the thumb URL you provided, it produces an error:

```
{"type": "Openslide", "error": "Unsupported or missing image file"}
```

**Comment by birm** [comment]
That's interesting.
so, I suspect this may have something to do with existing file names. Right now, if you ask the loader to write over a file, it refuses, and just calls it a bad request. (Same for other file operations, among other issues).

I guess, see if deleting the files (but not the /uploading folder itself) and trying again helps.

In the near future, we will work to make this process clearer, and support uploading files already dropped into that folder.

Please keep me posted,
-Ryan

**Comment by choosehappy** [comment]
sorry bit confused by your comment

it doesn't work if i use a previously unseen file, or an existing file

I've not yet had a file transfer go through successfully, regardless if it is a new or old file

for example, i just uploaded  PD10011a_HE-2014-12-07_15.02.40.svs

![image](https://user-images.githubusercontent.com/9681868/63616572-224bfa80-c5e8-11e9-829e-111b7240e40a.png)

clicking "check":

![image](https://user-images.githubusercontent.com/9681868/63616579-2841db80-c5e8-11e9-8553-e231b64f6a04.png)




looking at the output:

![image](https://user-images.githubusercontent.com/9681868/63616597-32fc7080-c5e8-11e9-8dbb-76deea7723fe.png)



and looking at the images directory:

![image](https://user-images.githubusercontent.com/9681868/63616614-3b54ab80-c5e8-11e9-9350-cbe3d2990995.png)


this is after deleting the contents of the ./image directory

does that help?

**Comment by birm** [comment]
Did you try to "finish" the upload in the middle?

**Comment by choosehappy** [comment]
yup, no dice

![image](https://user-images.githubusercontent.com/9681868/63617814-bc14a700-c5ea-11e9-8673-fc37ad98036f.png)
![image](https://user-images.githubusercontent.com/9681868/63617820-c0d95b00-c5ea-11e9-990c-f50d0e9dbbaf.png)


looks like less than half of the image made it up

similar proxy errors:

![image](https://user-images.githubusercontent.com/9681868/63617840-d2226780-c5ea-11e9-8232-02b9706f6ae7.png)

**Comment by birm** [comment]
this is certainly concerning. At the moment, I'm not sure exactly what is causing this issue. While I work on this, what exactly are you trying to accomplish? (e.g. Test the stack? Just post images yourself, or let anyone upload to the system?) I may be able to quickly put together a workaround that meets your needs.

Thanks again for your patience,
-Ryan

**Comment by choosehappy** [comment]
not a problem, i of course understand, technology can be a curse  :  )

most urgent on my todo list is to set up an online portal which will allow a handful of people to upload some WSIs and to share them round-about. 

we're executing a multi-site, multi-reader evaluation and this seemed like a good tool to do that in. correct? 

should be on the order of a few hundred slides

of course i'm also interested in the other functionalities, but playing with those are a bit lower on my to-do list at the moment

**Comment by birm** [comment]
Well, in this case we can't use a shortcut which avoids the "upload" part then...
Just to troubleshoot some common issues, can you tell me if there's enough space on the server on the disk/device that the distro folder is in?
And, when you start a file upload, it triggers a number of smaller posts with chunks of the files; do /any/ of these fail?

Thank you,
-Ryan

**Comment by choosehappy** [comment]
lots of space:

```
root@haeckel:~/Distro/images# df -h `pwd`
Filesystem                   Size  Used Avail Use% Mounted on
/dev/mapper/ubuntu--vg-root  498G  218G  273G  45% /
```

let me try deleting everything and starting from scratch

**Comment by choosehappy** [comment]
![image](https://user-images.githubusercontent.com/9681868/63620878-d5215600-c5f2-11e9-89e9-cedd86b2e24c.png)


so after clicking finish things looked okay, but then the check doesn't work


should i be clicking "finish upload" at any point after i select the file, or after the entirety of the file has been uploaded?

additionally, i pulled a previous version from the releases, and it had the same issue, so whatever it is might be local to these 2 machine's setup, or has existed for a long time

**Comment by choosehappy** [comment]
now on uploading a different file, i've clicked finish upload after selecting the file and adding the slidename, but things go off the rails quite quickly

![image](https://user-images.githubusercontent.com/9681868/63621082-537df800-c5f3-11e9-9635-51e81b22d0ac.png)

**Comment by birm** [comment]
finish upload should be once the entire file is done reading

**Comment by choosehappy** [comment]
so i've upgraded docker to the absolute latest version, and have waited until the UI says "Reading done" before clicking finish upload, and that appears to have helped:

![image](https://user-images.githubusercontent.com/9681868/63622596-790d0080-c5f7-11e9-9ae4-adc6d4d79511.png)

so now i click on "CAMICROSCOPE" in the landing page, i see:

"No Data Found ... x _ x"


but the files do look like they made it up:

```
root@haeckel:~/Distro/images# ls -ltr
total 117208
-rwxr-xr-x 1 root root  1938955 Aug 23 16:37 sample.svs
-rw-r--r-- 1 root root 73539609 Aug 23 16:39 16-01441.svs
-rw-r--r-- 1 root root 44527391 Aug 23 16:41 37571_001.svs
drwxr-xr-x 2 root root     4096 Aug 23 16:41 uploading
```

an error does appear at some point:

```
ca-security | public check true
ca-security | [HPM] Proxy created: /  ->  http://ca-front:80
ca-security | {"__statusCode":404,"err":"Not Found","type":"proxy error"}
ca-security | public check true
ca-security | [HPM] Proxy created: /  ->  http://ca-front:80
ca-security | {"__statusCode":404,"err":"Not Found","type":"proxy error"}

```

but not entirely sure when

is this helping at all?

**Comment by birm** [comment]
That error is probably something like a favicon or something like that.

Did you click "post" when you've finished the check?

**Comment by birm** [comment]
Whether or not this is the source of the issue, I agree that the instructions for this could be clearer or automated. thank you for pointing out this big flaw in our loader application.

**Comment by choosehappy** [comment]
: ) yes, this is a bit unintuitive, especially since it seems clicking any of the buttons at the wrong time brings everything to an irreversible halt

that said, so we 've built this website:

http://histoqcrepo.com/

which is open source:

https://github.com/foggydae/qcsliderepo-web

and has a _very_  nice user-friendly upload functionality that you may want to consider borrowing, drag and drop, progress bars, async uploading. the whole 9 yards. feel free to take the components and re-use

i'm still playing with this, but to my specific use case, if i have 5 pathologists, each with 100 slides that they want to upload, are they supposed to use this loader page and do them 1 by 1, or is there a UI bulk loading option? i saw a python script in one of the other repos, but i suspect they don't have the technical skills to execute something like that on their own

**Comment by birm** [comment]
You should be able to multiselect files in the file picker ui (i.e. shift+click), though yes not well described in the instructions.
We've actually been looking to use something like histoqc, but it's on hold at the moment. Would you be interested in working with us on this?

**Comment by choosehappy** [comment]
cool, multiselect does work, but i suspect it would be a bit of a burden if e.g., 100 slides were selected, as the user has to fill in the target "slidename" as i understand it? perhaps having it default to the uploading filename would already be a good option to help with bulk uploading

regarding histoqc, yes, of course, happy to collab. feel free to drop me an email when you're ready or if  you have any questions 

that said, i just want to point out that histoqc is different than histoqcrepo. histoqc is an open-source quality control tool, histoqcrepo.com is a publicly available website where users can upload and download slides which contain artifacts for their own testing purposes.

as a result of that we've had to build nice slide upload functionality using JS + html which you could potentially use to improve the uploader for caMicroscope.  it uses this JS library:

https://github.com/blueimp/jQuery-File-Upload

which is quite attractive. you can play with a demo of it here:

https://blueimp.github.io/jQuery-File-Upload/

![image](https://user-images.githubusercontent.com/9681868/63638296-b66c9f00-c686-11e9-9d1e-7d2506f3f7bc.png)

we ironed out whatever kinks were necessary for WSIs in our repo, and i can put you in touch with the developer who did it, to hopefully save you some grief : )

btw, thanks for your quick responses and development of this tool!
